[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/littilemike11/Csc-448-Spring-2022/blob/main/Michael_Edquilan_Assignment_4.ipynb)

In [ ]:
#author : Michael Edquilan

In [ ]:
import numpy as np
from sklearn import datasets # import iris dataset
from sklearn.linear_model import LogisticRegression 
from sklearn import svm # used for support vector machine
from sklearn.neural_network import MLPClassifier # import neural net
from sklearn.neighbors import KNeighborsClassifier # import KNN
from sklearn.model_selection import cross_val_score

In [ ]:
# import the Iris data set
iris = datasets.load_iris()
X = iris.data
Y = iris.target

In [ ]:
# setup Iris feature lablels
IrisF=("sepal length (cm)", "sepal width (cm)", "petal length (cm)", "petal width (cm)")
for i in range(0,4):
    print(IrisF[i])
for i in range(3):
    IrisC = ('Setosa','Versicolor','Virginica')
    print(IrisC[i])

sepal length (cm)
sepal width (cm)
petal length (cm)
petal width (cm)
Setosa
Versicolor
Virginica


In [ ]:
# split data into test(10%) and train (90%)
import numpy as np
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.10, random_state = 1)
X_train
Y_train
X_test
Y_test


array([0, 1, 1, 0, 2, 1, 2, 0, 0, 2, 1, 0, 2, 1, 1])

## Logistic Regression

In [ ]:
#LOGISTIC REGRESSION ORIGINAL
clf = LogisticRegression(random_state=0, max_iter= 10000)
clf.fit(X, Y)
clf.score(X, Y)

0.9733333333333334

In [ ]:
#LOGISTIC REGRESSION SPLIT
clf = LogisticRegression(random_state=0, max_iter = 10000)
clf.fit(X_train, Y_train)
clf.score(X_test, Y_test)

1.0

In [ ]:
#probability of sample data points
clf.predict_proba(X_test)

array([[9.86391118e-01, 1.36088748e-02, 7.13228564e-09],
       [2.78650159e-01, 7.19628828e-01, 1.72101311e-03],
       [5.76552450e-03, 9.07534112e-01, 8.67003631e-02],
       [9.86900514e-01, 1.30994649e-02, 2.14803367e-08],
       [8.55060810e-07, 1.89035893e-02, 9.81095556e-01],
       [4.17249158e-03, 6.87264310e-01, 3.08563198e-01],
       [1.23630754e-04, 1.21561379e-01, 8.78314990e-01],
       [9.58274395e-01, 4.17251455e-02, 4.59132966e-07],
       [9.71270277e-01, 2.87296087e-02, 1.14122304e-07],
       [7.46061719e-06, 2.56066327e-02, 9.74385907e-01],
       [1.73123087e-02, 9.06018738e-01, 7.66689532e-02],
       [9.68017798e-01, 3.19820408e-02, 1.61101107e-07],
       [3.42142048e-06, 2.82410444e-02, 9.71755534e-01],
       [6.50979935e-03, 8.39097123e-01, 1.54393077e-01],
       [6.41032431e-03, 7.91046458e-01, 2.02543217e-01]])

In [ ]:
coef =clf.coef_
coef
#each column is a feature, each row is the different class of iris

array([[-0.44138155,  0.85881902, -2.4551124 , -0.99800215],
       [ 0.55096767, -0.31907664, -0.17530765, -0.93548914],
       [-0.10958613, -0.53974238,  2.63042005,  1.93349129]])

In [ ]:
intercept=clf.intercept_
intercept
#each column is a seperate class

array([  9.94655972,   1.87091777, -11.8174775 ])

In [ ]:
#equation of different iris classes 
for i in range(3):
    print (IrisC[i]," : ", coef[i][0],"x1 + ", coef[i][1],"x2 + ", coef[i][2],"x3 + ", coef[i][3],"x4 + ", intercept[i])   

Setosa  :  -0.4413815450567822 x1 +  0.8588190168117873 x2 +  -2.4551123983410412 x3 +  -0.9980021522808156 x4 +  9.946559724066795
Versicolor  :  0.5509676721290918 x1 +  -0.3190766368082094 x2 +  -0.17530765268505535 x3 +  -0.9354891353279474 x4 +  1.8709177713103367
Virginica  :  -0.10958612707232307 x1 +  -0.5397423800035863 x2 +  2.630420051026092 x3 +  1.933491287608762 x4 +  -11.817477495377219


We got a score of 1.0 on the logistic regression model. This means that the model did a perfect job at classifying the different iris classes given the different feature values. The score measures the percent of times it correctly predicted a iris' class.

## Support Vector Machine

In [ ]:
# SUPPORT VECTOR MACHINE ORIGINAL
clf = svm.SVC()
clf.fit(X, Y)
clf.score(X, Y)

0.9733333333333334

In [ ]:
# SUPPORT VECTOR MACHINE SPLIT
clf = svm.SVC()
clf.probability=True
clf.fit(X_train, Y_train)
clf.score(X_test, Y_test)

1.0

In [ ]:
# prediction of sample data points
clf.predict_proba(X_test)

array([[0.97595187, 0.01396584, 0.01008229],
       [0.11040119, 0.86874662, 0.02085219],
       [0.00758756, 0.96771746, 0.02469499],
       [0.97438073, 0.01583952, 0.00977974],
       [0.01717648, 0.00482126, 0.97800226],
       [0.00844412, 0.80612238, 0.1854335 ],
       [0.01062964, 0.04064819, 0.94872217],
       [0.93766625, 0.04702299, 0.01531076],
       [0.96053072, 0.02845371, 0.01101556],
       [0.00802901, 0.00313669, 0.98883431],
       [0.00670412, 0.97858268, 0.0147132 ],
       [0.95255555, 0.03397369, 0.01347076],
       [0.00828946, 0.00275113, 0.98895941],
       [0.00752138, 0.93840971, 0.0540689 ],
       [0.00590193, 0.91663635, 0.07746172]])

We got a score of 1.0 on the support vector machine model. This means that the model did a perfect job at classifying the different iris classes given the different feature values. The score measures the percent of times it correctly predicted a iris' class. Meaning that it classified all the sample correctly.

## Neural Network

In [ ]:
# NEURAL NETWORK ORIGINAL
max_score = 0
maxi =0
maxj=0
for i in range(1,11): # test 
    for j in range(1,11):
        clf = MLPClassifier(solver='lbfgs', alpha=1e-4, hidden_layer_sizes=(i, j), random_state=0, max_iter = 10000)
        clf.fit(X,Y)
        score =clf.score(X,Y)
        if score > max_score:
            max_score = score
            maxi =i
            maxj =j
        if score > .95:
            print("Hidden layer size : (",i, ",", j, "), : score = ",score)                
print("Best Hidden layer size : (",maxi, ",", maxj, "), : score = ",max_score) 

Hidden layer size : ( 1 , 2 ), : score =  0.9933333333333333
Hidden layer size : ( 1 , 3 ), : score =  0.9866666666666667
Hidden layer size : ( 1 , 6 ), : score =  0.9933333333333333
Hidden layer size : ( 1 , 8 ), : score =  0.9866666666666667
Hidden layer size : ( 2 , 2 ), : score =  0.9866666666666667
Hidden layer size : ( 2 , 3 ), : score =  0.9866666666666667
Hidden layer size : ( 2 , 4 ), : score =  0.9866666666666667
Hidden layer size : ( 2 , 5 ), : score =  0.9866666666666667
Hidden layer size : ( 2 , 6 ), : score =  0.9866666666666667
Hidden layer size : ( 2 , 7 ), : score =  0.9866666666666667
Hidden layer size : ( 2 , 8 ), : score =  0.9866666666666667
Hidden layer size : ( 2 , 9 ), : score =  0.9866666666666667
Hidden layer size : ( 2 , 10 ), : score =  0.9866666666666667
Hidden layer size : ( 3 , 2 ), : score =  0.9866666666666667
Hidden layer size : ( 3 , 3 ), : score =  0.9866666666666667
Hidden layer size : ( 3 , 4 ), : score =  0.9866666666666667
Hidden layer size : ( 3

In [ ]:
# NEURAL NETWORK SPLIT
max_score = 0
maxi =0
maxj=0
for i in range(1,11): # test 
    for j in range(1,11):
        clf = MLPClassifier(solver='lbfgs', alpha=1e-4, hidden_layer_sizes=(i, j), random_state=0, max_iter = 10000)
        clf.fit(X_train,Y_train)
        score =clf.score(X_test,Y_test)
        if score > max_score:
            max_score = score
            maxi =i
            maxj =j
        if score > .95:
            print("Hidden layer size : (",i, ",", j, "), : score = ",score)    
print("Best Hidden layer size : (",maxi, ",", maxj, "), : score = ",max_score) 

Hidden layer size : ( 1 , 2 ), : score =  1.0
Hidden layer size : ( 1 , 3 ), : score =  1.0
Hidden layer size : ( 1 , 6 ), : score =  1.0
Hidden layer size : ( 1 , 8 ), : score =  1.0
Hidden layer size : ( 2 , 2 ), : score =  1.0
Hidden layer size : ( 2 , 3 ), : score =  1.0
Hidden layer size : ( 2 , 4 ), : score =  1.0
Hidden layer size : ( 2 , 5 ), : score =  1.0
Hidden layer size : ( 2 , 6 ), : score =  1.0
Hidden layer size : ( 2 , 7 ), : score =  1.0
Hidden layer size : ( 2 , 8 ), : score =  1.0
Hidden layer size : ( 2 , 9 ), : score =  1.0
Hidden layer size : ( 2 , 10 ), : score =  1.0
Hidden layer size : ( 3 , 2 ), : score =  1.0
Hidden layer size : ( 3 , 3 ), : score =  1.0
Hidden layer size : ( 3 , 4 ), : score =  1.0
Hidden layer size : ( 3 , 5 ), : score =  1.0
Hidden layer size : ( 3 , 6 ), : score =  1.0
Hidden layer size : ( 3 , 7 ), : score =  1.0
Hidden layer size : ( 3 , 8 ), : score =  1.0
Hidden layer size : ( 3 , 10 ), : score =  1.0
Hidden layer size : ( 4 , 2 ), :

In [ ]:
clf.predict_proba(X_test)

array([[1.00000000e+000, 3.00551256e-026, 1.32042901e-040],
       [9.99163249e-032, 9.99999735e-001, 2.65083783e-007],
       [2.89748523e-061, 9.99971699e-001, 2.83007119e-005],
       [1.00000000e+000, 1.71078719e-020, 2.96154387e-032],
       [1.49982880e-125, 9.09733686e-005, 9.99909027e-001],
       [1.27822685e-070, 9.98795942e-001, 1.20405783e-003],
       [2.54110045e-118, 3.87254670e-005, 9.99961275e-001],
       [1.00000000e+000, 1.49498258e-013, 2.62515583e-024],
       [1.00000000e+000, 1.91613081e-020, 9.06314850e-032],
       [3.93546879e-133, 3.07207521e-007, 9.99999693e-001],
       [4.47392559e-055, 9.99976331e-001, 2.36690786e-005],
       [1.00000000e+000, 1.52739485e-015, 1.94086747e-026],
       [1.06260700e-131, 8.54255083e-007, 9.99999146e-001],
       [4.89989520e-064, 9.99904708e-001, 9.52916435e-005],
       [2.47460593e-067, 9.99081939e-001, 9.18061380e-004]])

We got a score of 1.0 on the Neueral Network model. This means that the model did a perfect job at classifying the different iris classes given the different feature values. The score measures the percent of times it correctly predicted a iris' class.

## K Nearest Neighbor

In [ ]:
# K NEAREST NEIGHBOR ORINIAL
for i in range(1,11): # test knn with k ranging from 1-10
    clf = KNeighborsClassifier(n_neighbors=i)
    clf.fit(X, Y)
    print(i, " Nearest Neighbors : score =",clf.score(X, Y))

1  Nearest Neighbors : score = 1.0
2  Nearest Neighbors : score = 0.98
3  Nearest Neighbors : score = 0.96
4  Nearest Neighbors : score = 0.96
5  Nearest Neighbors : score = 0.9666666666666667
6  Nearest Neighbors : score = 0.9733333333333334
7  Nearest Neighbors : score = 0.9733333333333334
8  Nearest Neighbors : score = 0.98
9  Nearest Neighbors : score = 0.98
10  Nearest Neighbors : score = 0.98


In [ ]:
# K NEAREST NEIGHBOR SPLIT
for i in range(1,11): # test knn with k ranging from 1-10
    clf = KNeighborsClassifier(n_neighbors=i)
    clf.fit(X_train, Y_train)
    print(i, " Nearest Neighbors : score =",clf.score(X_test, Y_test))

1  Nearest Neighbors : score = 1.0
2  Nearest Neighbors : score = 1.0
3  Nearest Neighbors : score = 1.0
4  Nearest Neighbors : score = 1.0
5  Nearest Neighbors : score = 1.0
6  Nearest Neighbors : score = 1.0
7  Nearest Neighbors : score = 1.0
8  Nearest Neighbors : score = 1.0
9  Nearest Neighbors : score = 1.0
10  Nearest Neighbors : score = 1.0


In [ ]:
clf = KNeighborsClassifier(n_neighbors=1)
clf.fit(X_train, Y_train)
clf.predict_proba(X_test)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 1., 0.],
       [1., 0., 0.],
       [0., 0., 1.],
       [0., 1., 0.],
       [0., 0., 1.],
       [1., 0., 0.],
       [1., 0., 0.],
       [0., 0., 1.],
       [0., 1., 0.],
       [1., 0., 0.],
       [0., 0., 1.],
       [0., 1., 0.],
       [0., 1., 0.]])

We got a score of 1.0 on the K Nearest Nieghbor model with a value of k. This means that the model did a perfect job at classifying the different iris classes given the different feature values. The score measures the percent of times it correctly predicted a iris' class. The second best k value is 2 with a score .98 meaning it correctly classified the sample 98% of the time

## Conclution and Takeaways
I noticed that all of models performed well in classifying the iris dataset, as their scores were almost always perfect (1.0). I did find it all that the scores were so well, but this can be due to the fact that there isn't a lot of data in the set. Especially, when the data is split up 9:1 of training to test data, we are only predicting 15 samples, therefore it is more liekly to get better scores on smaller test sizes. As an alternative I also calculated the score based on the original dataset without any splitting. In regards to the split data(X_train, X_test, Y_train, Y_test), the best model is KNN where k =1 because it is the simplest model, in that it doesnt require "training" while still maintaining a perfect score. In regards to the oriniginal data ( the unsplit datat), the best model would also be KNN where k =1 as it has a perfect score. The second best option would be neural nets with hidden layer (1,2) as it has a score of .993333 